karapace/protobuf/proto_normalizations.py

```diff
@@ -0,0 +1,220 @@
+    sorted_fields = (
+        None if group.fields is None else [type_field_element_with_sorted_options(field) for field in group.fields]
+    )
+    return GroupElement(
+        label=group.label,
+        location=group.location,
+        name=group.name,
+        tag=group.tag,
+        documentation=group.documentation,
+        fields=sorted_fields,
+    )
+
+
+def one_ofs_with_sorted_options(one_ofs: OneOfElement) -> OneOfElement:
+    sorted_options = None if one_ofs.options is None else list(sorted(one_ofs.options, key=sort_by_name))
+    sorted_fields = [type_field_element_with_sorted_options(field) for field in one_ofs.fields]
+    sorted_groups = [groups_with_sorted_options(group) for group in one_ofs.groups]
+
+    return OneOfElement(
+        name=one_ofs.name,
+        documentation=one_ofs.documentation,
+        fields=sorted_fields,
+        groups=sorted_groups,
+        options=sorted_options,
+    )
+
+
+def message_element_with_sorted_options(message_element: MessageElement) -> MessageElement:
+    sorted_options = None if message_element.options is None else list(sorted(message_element.options, key=sort_by_name))
+    sorted_neasted_types = [type_element_with_sorted_options(nested_type) for nested_type in message_element.nested_types]
+    sorted_fields = [type_field_element_with_sorted_options(field) for field in message_element.fields]
+    sorted_one_ofs = [one_ofs_with_sorted_options(one_of) for one_of in message_element.one_ofs]
+
+    return MessageElement(
+        location=message_element.location,
+        name=message_element.name,
+        documentation=message_element.documentation,
+        nested_types=sorted_neasted_types,
+        options=sorted_options,
+        reserveds=message_element.reserveds,
+        fields=sorted_fields,
+        one_ofs=sorted_one_ofs,
+        extensions=message_element.extensions,
+        groups=message_element.groups,
+    )
+
+
+def type_element_with_sorted_options(type_element: TypeElement) -> TypeElement:
+    sorted_neasted_types: List[TypeElement] = []
+
+    for nested_type in type_element.nested_types:
+        if isinstance(nested_type, EnumElement):
+            sorted_neasted_types.append(enum_element_with_sorted_options(nested_type))
+        elif isinstance(nested_type, MessageElement):
+            sorted_neasted_types.append(message_element_with_sorted_options(nested_type))
+        else:
+            raise ValueError("Unknown type element")  # tried with assert_never but it did not work
+
+    # doing it here since the subtypes do not declare the nested_types property
+    type_element.nested_types = sorted_neasted_types
+
+    if isinstance(type_element, EnumElement):
+        return enum_element_with_sorted_options(type_element)
+
+    if isinstance(type_element, MessageElement):
+        return message_element_with_sorted_options(type_element)
+
+    raise ValueError("Unknown type element")  # tried with assert_never but it did not work
```

Comment:
@aiven-anton any suggestion on how I can tell `mypy` that all the concrete instances are managed? I've also tried with `assert type(type_element) != TypeElement` (since its a concrete class and could be instantiated) but didn't helped

Reply:
@eliax1996 The only way to define a hierarchy that works like ADTs in Python is to define base class and union separately. There's been proposals to introduce a concept of sealed classes but so far it has met a fair bit of justified critique, and it was [raised again recently](https://discuss.python.org/t/draft-pep-sealed-decorator-for-static-typing/49206/20) without much excitement.

What you can do currently is the below, and use `Thing` as type.

```python

class BaseThing: ...

class MemberA(BaseThing): ...
class MemberB(BaseThing): ...

Thing = MemberA | MemberB
```

The reason mypy rejects your code is it fails to prove there aren't any other subclasses of `TypeElement` it hasn't found yet. Theoretically this could be proven, but not with reasonable performance. You should also be able to see this with a `reveal_type(type_element)` inserted above the `raise ValueError`. I would expect the type to be `TypeElement` there.

Reply:
got it, I was thinking python was based a closed world model and so the classes where implicitly sealed. But isn't true, in fact you can always use a library and define a child of a class belonging to that library, so the checking isn't always exaustive